torch/_inductor/fx_passes/b2b_gemm.py

```diff
@@ -0,0 +1,188 @@
+import torch
+from ..._dynamo.utils import counters
+from ..ir import FixedLayout
+from ..pattern_matcher import Arg, CallFunction, Match, register_graph_pattern, PatternMatcherPass
+from ..select_algorithm import (
+    TritonTemplate,
+    autotune_select_algorithm,
+)
+from ..utils import ceildiv
+
+aten = torch.ops.aten
+
+# dram load/store estimations
+#   (A @ B) @ C
+#   M * N, N * O, O * P
+#   baseline
+#     load = M * N + N * O + M * O + O * P
+#     store = M * O + M * P
+#   gemm
+#     load = M * N + M / m * (N * O + O * P)
+#     store = M * P
+
+def b2b_gemm_grid(m, n, meta):
+    """
+    The CUDA grid size for matmul triton templates.
+    """
+    return (
+        ceildiv(m, meta["ROW_BLOCK_SIZE"]) * ceildiv(n, meta["COL_BLOCK_SIZE"]),
+        1,
+        1,
+    )
+
+b2b_gemm_template = TritonTemplate(
+    name="b2b_gemm",
+    grid=b2b_gemm_grid,
+    debug=True,
+    source=r"""
+{{def_kernel("A", "B", "C")}}
+    M = {{size("A", 0)}}
+    # N = {{size("A", 1)}}
+    O = {{size("C", 0)}}
+    # P = {{size("C", 1)}}
+
+    stride_am = {{stride("A", 0)}}
+    stride_an = {{stride("A", 1)}}
+    stride_bn = {{stride("B", 0)}}
+    stride_bo = {{stride("B", 1)}}
+    stride_co = {{stride("C", 0)}}
+    stride_cp = {{stride("C", 1)}}
+
+    # A's row block for this thread
+    row_block_id = tl.program_id(axis=0)
+
+    # divide B's columns (and C's rows)
+    num_col_block = tl.cdiv(O, COL_BLOCK_SIZE)
+
+    # offsets (TODO: handle the non-divisible case)
+    offs_row = row_block_id * ROW_BLOCK_SIZE + tl.arange(0, ROW_BLOCK_SIZE)
+    offs_col = tl.arange(0, COL_BLOCK_SIZE)  # to be updated in the loop
+
+    # accumulator
+    acc = tl.zeros((ROW_BLOCK_SIZE, P), dtype=tl.float16)
+
+    a_ptrs = A + (offs_row[:, None] * stride_am + tl.arange(0, N)[None, :] * stride_an)
+    a = tl.load(a_ptrs)
+
+    for _ in range(num_col_block):
+
+        b_ptrs = B + (tl.arange(0, N)[:, None] * stride_bn + offs_col[None, :] * stride_bo)
+        b = tl.load(b_ptrs)
+
+        c_ptrs = C + (offs_col[:, None] * stride_co + tl.arange(0, P)[None, :] * stride_cp)
+        c = tl.load(c_ptrs)
+
+        # computation (TODO: floating point errors)
+        acc += tl.dot(tl.dot(a, b, out_dtype=tl.float16), c, out_dtype=tl.float16)
+
+        # update offsets
+        offs_col += COL_BLOCK_SIZE
+
+    # store
+    idx_m = offs_row[:, None]
+    idx_p = tl.arange(0, P)[None, :]
+    mask = (idx_m < M) & (idx_p < P)
+
+    {{store_output(("idx_m", "idx_p"), "acc", "mask")}}
+""",
+)
+
+def can_apply_b2b_gemm(mat1, mat2, mat3) -> bool:
+    if not (("val" in mat1.meta) and ("val" in mat2.meta) and ("val" in mat3.meta)):
+        return False
+    mat1 = mat1.meta["val"]
```

Comment:
Purely cosmetic, but I wonder if the code would look a bit cleaner to do things like:

```
metas = [m.meta for m in [mat1, mat2, mat3]]
if not all("val" in meta for meta in metas):
    return False

mats = [m.meta["val'] for m in metas]
if not all(m.is_cuda for m in mats):
    return False
```

(Assuming I actually got that right). I'm not sure that looks better. Just a suggestion; up to you.

Reply:
Updated.